src/transformers/models/mplug_owl/modeling_mplug_owl.py

```diff
@@ -0,0 +1,1574 @@
+
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+
+        >>> prompt = [
+        ...     "The following is a conversation between a curious human and AI assistant. The assistant gives helpful, detailed, and polite answers to the user's questions.\nHuman: <image>\nHuman: how many cats are there?\nAI: "
+        ... ]
+        >>> inputs = processor(images=[image], text=prompt, return_tensors="pt").to(device, torch.float16)
+
+        >>> generated_ids = model.generate(**inputs)
+        >>> generated_text = processor.batch_decode(generated_ids, skip_special_tokens=True)[0].strip()
+        >>> generated_text
+        'There are two cats in the image.'
+        ```"""
+        if pixel_values is not None:
+            pixel_values = pixel_values.to(self.vision_model.embeddings.cls_token.data.dtype)
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+        # get text embedding
+        text_tokens_ = input_ids.clone()
+        batch_size = input_ids.shape[0]
+
+        media_token_indices = [get_media_indices(text_tokens_[i][:-1]) for i in range(batch_size)]
+        text_tokens_[text_tokens_ < 0] = 1  # Not used
+        text_embeds = self.get_input_embeddings()(text_tokens_)  # Temporally Embedding
+
+        if pixel_values is not None:
+            image_embeds = self.vision_model(pixel_values, return_dict=True).last_hidden_state
+
+            image_attention_mask = torch.ones(image_embeds.size()[:-1], dtype=torch.long, device=image_embeds.device)
+            query_tokens = self.query_tokens.expand(image_embeds.shape[0], -1, -1)
+
+            query_features = self.abstractor(
+                query_embeds=query_tokens,
+                encoder_hidden_states=image_embeds,
+                encoder_attention_mask=image_attention_mask,
+            )["last_hidden_state"]
+            torch.ones(query_features.size()[:-1], dtype=torch.long).to(query_features.device)
+            img_seq_length = query_features.shape[1]
+        if num_images is not None:
+            num_images_per_sample = num_images.long().cpu().tolist()
+        else:
+            num_images_per_sample = [0] * input_ids.shape[0]
+
+        text_chunk_embeds = []
+        img_idx = 0
+        for b in range(batch_size):
+            start = 0
+            result = []
+            if len(media_token_indices[b]) > 0:
+                for i, pos in enumerate(media_token_indices[b]):
+                    if pos > start:
+                        result.append(text_embeds[b, start:pos])
+                    result.append(query_features[img_idx + i])
+                    start = pos + img_seq_length
+            if start < text_embeds.shape[1]:
+                result.append(text_embeds[b, start:])
+
+            img_idx += num_images_per_sample[b]
+            text_chunk_embeds.append(torch.cat(result, dim=0))
```

Comment:
Can you please wrap this in a private method and commenting what the method does? also, can you use `batch` instead of batch size